### PR TITLE
Add `context` object to `GraphQLDataSource.didReceiveResponse` arguments

### PR DESCRIPTION
**Pull request by trevor-scheer**
This adds a third parameter `context` to the `didReceiveResponse` method on a `GraphQLDataSource`. This gives the hook access to the context object, allowing it to read and modify as necessary.

As demonstrated by the included test, this enables a `GraphQLDataSource` to inspect the `response` and add information to `context` accordingly.

This also provides parity with another hook, `willSendRequest`, which already has access to the `context` object.

### CODE REVIEW

File: packages/apollo-gateway/src/datasources/RemoteGraphQLDataSource.ts [hunk 1]
```diff
@@ -87,9 +91,10 @@ export class RemoteGraphQLDataSource implements GraphQLDataSource {
     >,
   ): ValueOrPromise<void>;
 
-  public async didReceiveResponse<TResult = any>(
+  public async didReceiveResponse<TContext, TResult = any>(
```
Comment (abernix) [inline packages/apollo-gateway/src/datasources/RemoteGraphQLDataSource.ts]: Fly-by comment: Adding this as the first positional type argument would be a breaking change for any existing implementors (who aren't us, since we just rely on the default of `any`) — can it be added after `TResult = any` with its own default value?

(Realistically though, we should probably have this `TContext` as a type variable on the entire class rather than having it localize to this method.)

Comment (trevor-scheer) [reply packages/apollo-gateway/src/datasources/RemoteGraphQLDataSource.ts]: You're absolutely right. I took the easy way in 0ecd617.

I'll take another stab at `TContext` on the class, but `RemoteGraphQLDataSource.process()` wasn't really having it.

File: docs/source/federation/implementing.md [hunk 2]
```diff
@@ -200,4 +200,64 @@ server.listen().then(({ url }) => {
 
 By leveraging the `buildService` function, we're able to customize how requests are sent to our federated services. In this example, we return a custom `RemoteGraphQLDataSource`. The datasource allows us to modify the outgoing request with information from the Apollo Server `context` before it's sent. Here, we're adding the `user-id` header to pass an authenticated user id to downstream services.
 
+In a similar vein, the `didReceiveResponse` hook allows us to inspect a service's `response` in order to modify the `context`. The lifecycle of a request to a federated server involves a number of responses, which may contain headers that need to be passed back to the client. Suppose we want to collect the <a href="https://docs.fastly.com/en/guides/getting-started-with-surrogate-keys" target="_blank" rel="noopener noreferrer">Surrogate-Key</a> headers from every response for caching purposes. We can implement this using the `didReceiveResponse` hook and an `ApolloServerPlugin`:
+
+```javascript{11-22,39-48}
```
Comment (trevor-scheer) [inline docs/source/federation/implementing.md]: Fix line numbers
```suggestion
```javascript{12-23,40-49}
```

File: docs/source/federation/implementing.md [hunk 3]
```diff
@@ -200,4 +200,64 @@ server.listen().then(({ url }) => {
...
+      didReceiveResponse(response, request, context) {
+        if (response.ok) {
+          // Parse surrogate keys from the header and push them onto the context
+          const surrogateKeyString = response.headers.get('Surrogate-Key');
+          if (surrogateKeyString) {
+            context.surrogateKeys.push(...surrogateKeyString.split(' '));
+          }
+          return this.parseBody(response);
+        } else {
+          throw await this.errorFromResponse(response);
```
Comment (abernix) [inline docs/source/federation/implementing.md]: Hmm, this `await` is inside of a non-`async` function.

File: packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts [hunk 4]
```diff
@@ -115,6 +119,50 @@ describe('willSendRequest', () => {
   });
 });
 
+describe('didReceiveResponse', () => {
```
Comment (abernix) [inline packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts]: Can we add another test that tests the use of `didReceiveResponse` without the `context` parameter being passed (the previous behavior)?

File: packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts [hunk 5]
```diff
@@ -115,6 +119,50 @@ describe('willSendRequest', () => {
...
+  it('can accept and modify context', async () => {
+    interface MyContext {
+      surrogateKeys: string[];
+    }
+
+    const DataSource = new RemoteGraphQLDataSource({
+      url: 'https://api.example.com/foo',
+      async didReceiveResponse(response, request, context) {
+        // I'd prefer to use super.didReceiveResponse here but it doesn't seem to work.
+        const body = await RemoteGraphQLDataSource.prototype.didReceiveResponse.call(
```
Comment (abernix) [inline packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts]: Aha, `super` doesn't work because there is no super class of `RemoteGraphQLDataSource` when overriding the method from the constructor in the way that we area; `RemoteGraphQLDataSource` isn't an extension of anything, it only `implements GraphQLDataSource`:

https://github.com/apollographql/apollo-server/blob/e6ccd8bc41e51089ca428c408be2bcd0a8e3796d/packages/apollo-gateway/src/datasources/RemoteGraphQLDataSource.ts#L21

(So no `super`class!)

File: packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts [hunk 6]
```diff
@@ -115,6 +119,50 @@ describe('willSendRequest', () => {
...
+        );
+        const surrogateKeys = request.headers.get('surrogate-keys');
+        if (surrogateKeys) {
+          (context as any).surrogateKeys.push(...surrogateKeys.split(' '));
+        }
+        return body;
+      },
+    });
+
+    fetch.mockJSONResponseOnce({ data: { me: 'james' } });
```
Comment (abernix) [inline packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts]: This test should fail.  You are not `james`.

:trollface:

File: packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts [hunk 7]
```diff
@@ -125,6 +173,7 @@ describe('error handling', () => {
 
     const result = DataSource.process({
       request: { query: '{ me { name } }' },
+      context: {},
```
Comment (abernix) [inline packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts]: Should we continue to test the existing case where `context` was not present?

File: packages/apollo-datasource-rest/src/RESTDataSource.ts [hunk 8]
```diff
@@ -95,6 +95,7 @@ export abstract class RESTDataSource<TContext = any> extends DataSource {
   protected async didReceiveResponse<TResult = any>(
     response: Response,
     _request: Request,
+    _context: TContext,
```
Comment (abernix) [inline packages/apollo-datasource-rest/src/RESTDataSource.ts]: It's great that `TContext` already exists here!

I think this should be an optional parameter to account for the existing pattern?
```suggestion
    _context?: TContext,
```

File: packages/apollo-datasource-rest/src/RESTDataSource.ts [hunk 9]
```diff
@@ -258,7 +259,7 @@ export abstract class RESTDataSource<TContext = any> extends DataSource {
             cacheKey,
             cacheOptions,
           });
-          return await this.didReceiveResponse(response, request);
+          return await this.didReceiveResponse(response, request, this.context);
```
Comment (abernix) [inline packages/apollo-datasource-rest/src/RESTDataSource.ts]: Hmm, feel free to back out the changes to `RESTDataSource` if this is going too far, but I didn't realize that `this.context` was already on the instance.  Therefore, passing it into `didReceiveResponse` seems seem unnecessary since it can just be accessed from the `didReceiveResponse` method as `this.context` already!

File: docs/source/federation/implementing.md [hunk 10]
```diff
@@ -200,4 +200,64 @@ server.listen().then(({ url }) => {
 
 By leveraging the `buildService` function, we're able to customize how requests are sent to our federated services. In this example, we return a custom `RemoteGraphQLDataSource`. The datasource allows us to modify the outgoing request with information from the Apollo Server `context` before it's sent. Here, we're adding the `user-id` header to pass an authenticated user id to downstream services.
 
+In a similar vein, the `didReceiveResponse` hook allows us to inspect a service's `response` in order to modify the `context`. The lifecycle of a request to a federated server involves a number of responses, which may contain headers that need to be passed back to the client. Suppose we want to collect the <a href="https://docs.fastly.com/en/guides/getting-started-with-surrogate-keys" target="_blank" rel="noopener noreferrer">Surrogate-Key</a> headers from every response for caching purposes. We can implement this using the `didReceiveResponse` hook and an `ApolloServerPlugin`:
```
[the remark this answers is not in the document]
Comment (trevor-scheer) [reply docs/source/federation/implementing.md]: I think generalizing this example is a good suggestion.

As a side note, I'm curious if converting this file to `.mdx` would resolve the italics issue, though this isn't immediately relevant anymore if I drop the example.

[the remark this answers is not in the document]
Comment (trevorblades) [reply docs/source/federation/implementing.md]: @abernix yes! We can definitely apply `target="_blank" rel="noopener noreferrer"` to all external links in markdown by default. If that's what you're referring to, I can implement that this week 👍

File: packages/apollo-datasource-rest/src/RESTDataSource.ts (hunk 9 above)
Comment (trevor-scheer) [reply packages/apollo-datasource-rest/src/RESTDataSource.ts]: Yeah, agreed.
fa80407

File: packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts (hunk 5 above)
Comment (trevor-scheer) [reply packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts]: Very helpful, thanks for pointing this out!
Resolved in 2792bdb

File: packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts (hunk 6 above)
Comment (trevor-scheer) [reply packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts]: Alas, we can only aspire to be.

File: docs/source/federation/implementing.md (hunk 3 above)
Comment (trevor-scheer) [reply docs/source/federation/implementing.md]: Great catch, and I'm curious why my linter didn't since I _thought_ that I linted this code.
Resolved in 171a101

File: docs/source/federation/implementing.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (trevor-scheer) [reply docs/source/federation/implementing.md]: @trevorblades that would be an awesome idea! The suggestion to do so was actually from @StephenBarlow, who recommended we open a new window for external links as opposed to navigating away from the docs. Enforcing this throughout would be wonderful!

[the remark this answers is not in the document]
Comment (trevor-scheer) [reply docs/source/federation/implementing.md]: @abernix I've updated with a new example in 171a101. I'd love to hear your thoughts. If you have any better examples in mind which feel like a natural fit for this, I'd be very open to them!

File: docs/source/federation/implementing.md [hunk 11]
```diff
@@ -200,4 +200,64 @@ server.listen().then(({ url }) => {
...
+          const surrogateKeyString = response.headers.get('Surrogate-Key');
+          if (surrogateKeyString) {
+            context.surrogateKeys.push(...surrogateKeyString.split(' '));
+          }
+          return this.parseBody(response);
+        } else {
+          throw await this.errorFromResponse(response);
+        }
+      },
+    });
```
[the remark this answers is not in the document]
Comment (trevor-scheer) [reply docs/source/federation/implementing.md]: Agree with 1 as the best option here!
Resolved in 171a101

File: packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts (hunk 7 above)
Comment (trevor-scheer) [reply packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts]: The addition of these empty objects is in response to TS errors that exist on master. Including these changes within this PR may be incorrect, but I do think they should be here.

The outstanding issue is that context is a _required_ field according to the types. What I'm curious to know is, since we previously guaranteed the _runtime_ validity of a missing `context` entry in our tests, should we continue to do so? I'm tempted to say no, as this seems like a pandora's box of runtime null/undefined expectations. The core idea being: just because it works, doesn't mean it's expected behavior.

I'd like to hear your thoughts, and happy to make these changes in a separate PR to prevent conflating the issue!

File: packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts (hunk 4 above)
Comment (trevor-scheer) [reply packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts]: See related: https://github.com/apollographql/apollo-server/pull/3360/files/968c6b8c5894de08d5c6801b3f7385ebd300695f#r330680535

File: packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts (hunk 7 above)
Comment (abernix) [reply packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts]: > I'm tempted to say no, as this seems like a pandora's box of runtime null/undefined expectations.

Not everyone is using TypeScript though, so I do believe there is value in testing exactly this.

For speed, I'm okay with leaving them here, though if this PR is squashed, it will be hard to identify why these were changed later. 😉

**Comment by jhampton** [comment]
[Flowchart for discussion](https://bit.ly/2V92rm2) @trevor-scheer and @abernix

File: docs/source/federation/implementing.md [hunk 12]
```diff
@@ -200,4 +200,66 @@ server.listen().then(({ url }) => {
...
+    const cacheControl = response.headers.get('Cache-Control');
+    if (cacheControl) {
+      const result = cacheControl.match(/max-age=(\d*)/);
+      if (Array.isArray(result) && result[1]) {
+        context.cacheControl.maxAge = context.cacheControl.maxAge
+          ? Math.min(context.cacheControl.maxAge, Number(result[1]))
+          : Number(result[1]);
+      }
+    }
+    return body;
```
Comment (abernix) [inline docs/source/federation/implementing.md]: I do like this example and think it's much better than surrogate keys.  However, I think we should just tone this town to be a generic example representing a number (e.g.).

More technically, because cache control is "complicated" and there is extraordinary variation ([_`Vary`'d_ literally, sometimes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Vary)!) in how such a header is interpreted (plus, `public`/`private` rules, etc.).  For example, if the downstream service isn't present at all, it could be argued that that represents a `max-age` of `0`, which wouldn't work in the above example.  Ultimately, I worry that this would be an endorsement for its use.

But just-for-fun editorial on this suggestion:

Of Note:

* `\d*` should be `\d+` as there _must_ be at least one digit.  (`*` means _0 or more_, whereas `+` is _1 or more_).
* I default to `0` when `Cache-control` is absent.
* Shorter, but I think same result.

```js
class DataSourceWithCacheControl extends RemoteGraphQLDataSource {
  async didReceiveResponse(response, request, context) {
    // Parse the Cache-Control header and update the value on context if it's
    // _less_ than the current value (or if no value is currently set).
    const [ , seconds = 0] = (response.headers.get('Cache-Control') || "").match(/max-age=(\d+)/) || [];
    context.cacheControl.maxAge = Math.min(context.cacheControl.maxAge || +seconds, +seconds);
    // Important: Invoke the super method to return the response body.
    return await super.didReceiveResponse(response, request, context);
  }
}
```

But again, I don't think we necessarily want a `cache-control` example because of the overall technical complications of interpreting it properly with such a small amount of code.

File: docs/source/federation/implementing.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (abernix) [reply docs/source/federation/implementing.md]: Replied! https://github.com/apollographql/apollo-server/pull/3360/files#r331118070

Also, opened an issue for @trevorblades to track the higher-level/global fix. :smile:

I think _this_ thread can be resolved in lieu of the new comment?

File: packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts (hunk 7 above)
Comment (trevor-scheer) [reply packages/apollo-gateway/src/datasources/__tests__/RemoteGraphQLDataSource.test.ts]: I separated this commit specifically for the sake of revertibility, so consider it done :)

I agree, and feel that your consideration here errs on the side of caution, so I'll revert for now and reconsider at a later time. I think there's some philosophical discussion to be had w.r.t. un/expected behavior, documentation, and types. I definitely understand the "not everyone uses TypeScript" argument, but there may be some nuance here for me to understand better.

File: docs/source/federation/implementing.md (hunk 12 above)
Comment (trevor-scheer) [reply docs/source/federation/implementing.md]: Resolved in 8898ad41e98722b3960b029984506a15b2336238

File: docs/source/federation/implementing.md [hunk 13]
```diff
@@ -200,60 +199,63 @@ server.listen().then(({ url }) => {
...
 
-In a similar vein, the `didReceiveResponse` hook allows us to inspect a service's `response` in order to modify the `context`. The lifecycle of a request to a federated server involves a number of responses, which may contain headers that need to be passed back to the client. Suppose our services use the <a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control" target="_blank" stop="_italicizing" rel="noopener noreferrer">Cache-Control</a> header's `max-age` value. We may have receive a number of these values from our services but we can only respond with one, which should be the minimum of all of the values. We can implement this using the `didReceiveResponse` hook and an `ApolloServerPlugin`:
+In a similar vein, the `didReceiveResponse` hook allows us to inspect a service's `response` in order to modify the `context`. The lifecycle of a request to a federated server involves a number of responses, which may contain headers that need to be passed back to the client. Suppose our services report back their id via a `Service-Id` header.
```
Comment (abernix) [inline docs/source/federation/implementing.md]: The example below shows `Server-Id`, but this text uses `Service-Id`.

File: docs/source/federation/implementing.md [hunk 14]
```diff
@@ -151,36 +151,35 @@ On startup, the gateway will fetch the service capabilities from the running ser
...
+  }
+}
+// highlight-end
+
 const gateway = new ApolloGateway({
   serviceList: [
     { name: 'products', url: 'http://localhost:4001' },
     // other services
   ],
+  // highlight-start
```
Comment (abernix) [inline docs/source/federation/implementing.md]: Cool! This is so much more manageable than line numbers!

File: docs/source/federation/implementing.md [hunk 15]
```diff
@@ -200,60 +199,63 @@ server.listen().then(({ url }) => {
...
+In a similar vein, the `didReceiveResponse` hook allows us to inspect a service's `response` in order to modify the `context`. The lifecycle of a request to a federated server involves a number of responses, which may contain headers that need to be passed back to the client. Suppose our services report back their id via a `Service-Id` header.
 
-```javascript{4-20,48,59-71}
+![Flowchart demonstrating willSendResponse usage](../images/willSendResponse-flowchart.png)
+
+To implement this, we can aggregate the service IDs into a single, comma-separated list in our final response using the `didReceiveResponse` hook and an `ApolloServerPlugin`:
```
Comment (abernix) [inline docs/source/federation/implementing.md]: This also says `service IDs`, while the example is `Server-Id`.  I think `Server-Id` is great and either is fine.

File: docs/source/federation/implementing.md (hunk 14 above)
Comment (trevor-scheer) [reply docs/source/federation/implementing.md]: Indeed, counting and updating the line numbers on this repeatedly was getting under my skin. These are long examples!

File: docs/source/federation/implementing.md (hunk 13 above)
Comment (trevor-scheer) [reply docs/source/federation/implementing.md]: Great catch! Resolved in 0171c44

File: docs/source/federation/implementing.md (hunk 15 above)
Comment (trevor-scheer) [reply docs/source/federation/implementing.md]: Resolved in 0171c44

**Comment by rtymchyk** [comment]
Excuse my inability to understand this, but I'm having trouble following what 'context' is here. Flow chart implies that it's the context of the final response, and some of the documentation has hints of that as well, but I wanted to confirm this. 

What I'd really like to do is pass through some response headers from service X to the final response headers of the gateway.
